iphone/Maps/Core/InappPurchase/InAppBilling.swift

```diff
@@ -0,0 +1,112 @@
+fileprivate struct BillingProduct: IBillingProduct {
+  var productId: String {
+    return product.productIdentifier
+  }
+
+  var localizedName: String {
+    return product.localizedTitle
+  }
+
+  var price: NSDecimalNumber {
+    return product.price
+  }
+
+  var priceLocale: Locale {
+    return product.priceLocale
+  }
+
+  let product: SKProduct
+
+  init(_ product: SKProduct) {
+    self.product = product
+  }
+}
+
+final class InAppBilling: NSObject, IInAppBilling {
+  private var productsCompletion: ProductsCompletion?
+  private var paymentCompletion: PaymentCompletion?
+  private var productRequest: SKProductsRequest?
+  private var billingProduct: BillingProduct?
+  private var pendingTransaction: SKPaymentTransaction?
+
+  override init() {
+    super.init()
+    SKPaymentQueue.default().add(self)
+  }
+
+  deinit {
+    SKPaymentQueue.default().remove(self)
+  }
+
+  func requestProduct(_ productIds: Set<String>, completion: @escaping ProductsCompletion) {
+    productsCompletion = completion
+    productRequest = SKProductsRequest(productIdentifiers: productIds)
+    productRequest!.delegate = self
+    productRequest!.start()
```

Comment:
According to SKProductsRequest documentation:

> Be sure to keep a strong reference to the request object; otherwise, the system might deallocate the request before it can complete.